presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1403,17 +1373,31 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
 
         if (handle.getBucketProperty().isPresent()) {
             ImmutableList<PartitionUpdate> partitionUpdatesForMissingBuckets = computePartitionUpdatesForMissingBuckets(session, handle, table.get(), partitionUpdates);
-            // replace partitionUpdates before creating the empty files so that those files will be cleaned up if we end up rollback
+            // replace partitionUpdates before creating the zero-row files so that those files will be cleaned up if we end up rollback
             partitionUpdates = PartitionUpdate.mergePartitionUpdates(Iterables.concat(partitionUpdates, partitionUpdatesForMissingBuckets));
+            HdfsContext hdfsContext = new HdfsContext(session, table.get().getDatabaseName(), table.get().getTableName());
             for (PartitionUpdate partitionUpdate : partitionUpdatesForMissingBuckets) {
                 Optional<Partition> partition = table.get().getPartitionColumns().isEmpty() ? Optional.empty() : Optional.of(buildPartitionObject(session, table.get(), partitionUpdate));
-                createEmptyFile(
+
+                Properties schema;
+                StorageFormat format;
+                if (partition.isPresent()) {
```

Comment:
Have we considered extracting this into a separate private method since it seems very similar to the one in `beginCreateTable()`?

Reply:
How about having two methods:

`Properties getHiveSchema(Table, Optional<Partition>)` and `StorageFormat getStorageFormat(Table, Optional<Partition>)`  

and use them inline?

e.g.:

```
zeroRowFileCreator.createFiles(
                        session,
                        session,
                        hdfsContext,
                        partitionUpdate.getWritePath(),
                        getTargetFileNames(partitionUpdate.getFileWriteInfos()),
                        getStorageFormat(table, partition),
                        handle.getCompressionCodec(),
                       getHiveSchema(table, partition));
```

Reply:
Sounds reasonable